### ISSUE TITLE
'self' passed to function decorated with @model_validator(mode="after") does not include PrivateAttr

### ISSUE THREAD

**Issue by RS-Credentive** [pydantic/pydantic#7142]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Related to #7083 

I believe this is the root of the failure, but I have not been able to figure out why it is occurring. I am providing a simpler code example to illustrate the problem:

Output of the code is:

```
Creating PrivateBase
dumping PrivateBase values - should be empty
Printing _private:
self._private is empty
Creating PrivateSub
Calling dump_private_attr from a validator
Printing _private:
self._private is empty
dumping PrivateSub values - should be 'PrivateSub'
Printing _private:
PrivateSub
```

### Example Code

```Python
from __future__ import annotations

from pydantic import BaseModel, PrivateAttr, model_validator

from typing import Any


class PrivateBase(BaseModel):
    _private: list[str] = PrivateAttr(default=[])

    def dump_private_attrs(self) -> PrivateBase:
        print("Printing _private:")
        if len(self._private) < 1:
            print("self._private is empty")
        for attr in self._private:
            print(attr)
        return self


class PrivateSub(PrivateBase):
    def __init__(self, **data: Any):
        super().__init__(**data)
        self._private.append("PrivateSub")

    @model_validator(mode="after")
    def validator_dump_private(self) -> PrivateSub:
        print("Calling dump_private_attr from a validator")
        self.dump_private_attrs()
        return self


print("Creating PrivateBase")
base = PrivateBase()
print("dumping PrivateBase values - should be empty")
base.dump_private_attrs()

print("Creating PrivateSub")
sub = PrivateSub()
print("dumping PrivateSub values - should be 'PrivateSub'")
sub.dump_private_attrs()
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.1.1
        pydantic-core version: 2.4.0
          pydantic-core build: profile=release pgo=false mimalloc=true
                 install path: /home/vscode/.local/lib/python3.11/site-packages/pydantic
               python version: 3.11.4 (main, Jun  7 2023, 19:03:45) [GCC 10.2.1 20210110]
                     platform: Linux-5.15.49-linuxkit-pr-aarch64-with-glibc2.31
     optional deps. installed: ['email-validator', 'typing-extensions']
```

**Comment by RS-Credentive** [comment]
A clue here, but I'm not sure why private attributes are in "model_post_init" instead of "model_init". Is that intentional?

https://github.com/pydantic/pydantic/blob/0454fabbc116ea78df0e29756a122cdbeba631d8/pydantic/_internal/_model_construction.py#L115

**Comment by dmontagu** [comment]
I think the reason was that adding complexity to `__init__` would have performance implications for everything, whereas adding a `model_post_init` is a way to more easily make it so that there is no performance overhead for models that don't require the function.

I wouldn't be surprised if there was a way to make this all work with just a single `__init__` and not have performance overhead for models without private attributes, but I'm not sure. @samuelcolvin might have more insight into why the `model_post_init` function is necessary.

**Comment by dmontagu** [comment]
CC @adriangb — interested if you have any idea about if there's a way to get private attributes to be accessible from an `@model_validator(mode='after')` given that private attributes get initialized after the call to `__init__`. I'm not sure how much of a problem it would be to use a modified `__init__`.

In general, I'd guess that a PR fixing this would be welcome if it didn't add overhead for non-users, and didn't introduce breaking changes. But I would guess that might be hard.

**Comment by RS-Credentive** [comment]
> I think the reason was that adding complexity to `__init__` would have performance implications for everything, whereas adding a `model_post_init` is a way to more easily make it so that there is no performance overhead for models that don't require the function.

Would there still be a performance hit if no private attributes were defined on a model? 

I will look into model_post_init for my use case. I assumed the validation would have been completed before model_post_init fired, but I might be wrong.

**Comment by RS-Credentive** [comment]
My ultimate goal is to do something similar to Validation Context here, but I don't want allowed values to be passed at run time. I want them to be defined on the model. This will allow me to create subclasses that include validation info in private attributes - which will make it very easy for third-party developers to define their own subclasses with validation info. For full context, my ultimate goal is something like this: https://github.com/RS-Credentive/oscal-pydantic/blob/oscal-pydantic-v2/src/oscal_pydantic/core/properties.py

The BaseProperty defines the basic model fields and some core constraints on them, and then subclasses allow you to specify different types of properties with different validation rules for different purposes. The validation rules in the standard I am following, [OSCAL](https://pages.nist.gov/OSCAL/), get pretty hairy.

**Comment by RS-Credentive** [comment]
After a bunch of debugging, I have a theory about what is failing.

When I create a subclass with a validator and then call `super().__init__()` from the subclass, the validator function is being passed in to the superclass as `__pydantic_validator__`. Since the superclass is generic and doesn't define the right values for the private attributes, the validator is failing. However, pydantic doesn't try the validator again within the context of the subclass; it just raises the error up the stack. 

Essentially, the subclass validator is failing in the superclass, and the exception is not being handled within the subclass, but is just raised up again.

Does my explanation make sense?

Of course, this may be the intended behavior. I might expect that the validator would be associated with the subclass, and that if I call `super().__init()`, the validator would be ignored in the context of the superclass (which might have its own validator). I think this would fix the problem, but I would welcome any corrections or feedback.

**Comment by RS-Credentive** [comment]
Definitely inelegant, but I was able to pass a type argument into a validation function (not decorated), and then add this to the function:

```
            if type(self) == calling_type:
                raise ValueError(
                    self.print_validation_errors(
                        validation_errors=self.validation_errors()
                    )
                )
            else:
                # This may be called from a subclass. If so, return control up
                return self
```

at each layer of validation, the function adds to an error list, and then checks to see if the validation function is being currently executed by the original caller.  If not, it passes control up the chain. If there are still errors when the original caller executes the function, then the error is raised.

I'm sure there's a better way, but that was something that helped me.

**Comment by RS-Credentive** [comment]
I have recorded a screenshare highlighting what I think the issue is. 

https://vimeo.com/855705403/0a2511aa10

**Comment by RS-Credentive** [comment]
I have recorded a screenshare that highlights the issue, I think:

https://vimeo.com/855705403/0a2511aa10

Note that I have changed the code slightly from what I posted to the original issue - see code listing below.

Let me know if I should open a new issue for this.

```py
class PrivateBase(BaseModel):
    _private: list[str] = PrivateAttr(default=[])

    def dump_private_attrs(self) -> typing.Self:
        print("Printing _private:")
        if len(self._private) < 1:
            print("self._private believes it is a " + str(type(self)))
            print("self._private is empty")
        for attr in self._private:
            print(attr)
        return self

    @model_validator(mode="after")
    def dump_from_validator(self) -> typing.Self:
        print("dump called from validator")
        return self.dump_private_attrs()


class PrivateSub(PrivateBase):
    def __init__(self, **data: typing.Any):
        super().__init__(**data)
        self._private.append("PrivateSub")


print("Creating PrivateBase")
base = PrivateBase()
print("dumping PrivateBase values - should be empty")
base.dump_private_attrs()

print("Creating PrivateSub")
sub = PrivateSub()
print("dumping PrivateSub values - should be 'PrivateSub'")
sub.dump_private_attrs()
```

**Comment by martinezpl** [comment]
This is ridiculous

**Comment by RS-Credentive** [comment]
Agreed. I have stopped using pydantic for new projects because of issues like this.

**Comment by axelmukwena** [comment]
Having this same issue!
